This notebook focuses on Multiple Linear Regression. 
The machine learning model will predict the price of the 2 house:
    One with area of 3000 sq. feet, 3 bedrooms and which is 40 years old in NJ, USA
    Second with area of 2500 sq. feet, 4 bedrooms, and which is 5 years old in NJ, USA.
Formula:
y = mx+c
price = m1*area + m2*bedrooms + m3*age + b
where, price is dependent variable and area, bedrooms and age are independent variables or features.
m1, m2 and m3 are coefficients and b is intercept

I will be performing:
1. Data preprocessing: Handling NA values
2. Linear Regression using Multipe Variables

In [1]:
import pandas as pd
import numpy as np
from sklearn import linear_model

In [4]:
df = pd.read_csv("C:/Users/shiml/Documents/shruti/Study/ML/Multiple Linear Regression/homeprices.csv")
df

area  bedroom  age   price
0  2600      3.0   20  550000
1  3000      4.0   15  565000
2  3200      NaN   18  610000
3  3600      3.0   30  595000
4  4000      5.0    8  760000

In [8]:
# Assumption: Taking median of all the values in bedroom to fill the NaN value
import math
bedroom_median = math.floor(df.bedroom.median())
bedroom_median

3

In [13]:
df.bedroom = df.bedroom.fillna(bedroom_median)
df

area  bedroom  age   price
0  2600      3.0   20  550000
1  3000      4.0   15  565000
2  3200      3.0   18  610000
3  3600      3.0   30  595000
4  4000      5.0    8  760000

In [14]:
reg = linear_model.LinearRegression()
reg.fit(df[['area', 'bedroom', 'age']], df.price)

LinearRegression()

In [16]:
reg.coef_

array([   137.25, -26025.  ,  -6825.  ])

In [17]:
reg.intercept_

383724.99999999994

In [19]:
reg.predict([[3000,3,40]])

c:\users\shiml\appdata\local\programs\python\python37\lib\site-packages\sklearn\base.py:451: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  "X does not have valid feature names, but"


array([444400.])

We can verify our prediction by cross-checking the dataset. In the dataset, we can see that a house of 3000 sq. feet, 4 bedrooms and 15 years old costs 565000. Where as here, the house of same area which has 1 less room and which is much older (40 years) costs less. Hence, 444400.

We can a linear relationship between the features and the dependent variable.
More the area of the house, more is costs.
More number of rooms, more it costs.
Older house costs less.

In [21]:
137.25*3000 + -26025*3 + -6825*40 + 383724.99999999994

444399.99999999994

In [22]:
reg.predict([[2500,4,5]])

c:\users\shiml\appdata\local\programs\python\python37\lib\site-packages\sklearn\base.py:451: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  "X does not have valid feature names, but"


array([588625.])

Similarly, in the data set there is a house which is 2600 sq. feet, has 3 bedrooms but is 20 years old. It costs 550000. Whereas this house costs, 2500 sq. feet, 4 bedrooms, and is much newer (5 years) costs more i.e. 588625